### PR TITLE
google ads pipeline

### PR DESCRIPTION
**Pull request by redicane**
# Tell us what you do here

- [] contributing new pipeline (please create an issue [here]() and link it below)
- [x] implementing verified pipeline (please link a relevant issue labelled as `verified pipeline`)
- [ ] fixing a bug (please link a relevant bug report)
- [ ] improving, documenting, customizing existing pipeline (please link relevant issue)
- [ ] anything else (please link an issue or describe below)

#87  Relevant issue

### CODE REVIEW

File: pipelines/google_ads/__init__.py [hunk 1]
```diff
@@ -0,0 +1,141 @@
+from typing import Dict, List, Union
+import dlt
+from dlt.common.configuration.specs import GcpClientCredentialsWithDefault
+from pipelines.google_analytics.helpers.credentials import GoogleAnalyticsCredentialsOAuth
```
Comment (rudolfix) [inline pipelines/google_ads/__init__.py:4]: please do not import cross-pipelines. we have good support for OAUTH now. Import `GcpOAuthCredentialsWithDefault` from specs. it is similar to yours

File: pipelines/google_ads/__init__.py [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+    :param credentials:
+    :param dev_token:
+    :param query_list:
+    :return:
+    """
+
+    # generate access token for credentials if we are using OAuth2.0
+    if isinstance(credentials, GoogleAnalyticsCredentialsOAuth):
+        credentials.auth("https://www.googleapis.com/auth/analytics.readonly")
+        credentials = {
```
Comment (rudolfix) [inline pipelines/google_ads/__init__.py:38]: you can just use `to_native_representation` if you import from our specs (I hope - if not then tell me)

Comment (rudolfix) [reply pipelines/google_ads/__init__.py:38]: or just use `to_google_credentials` that will return google credentials object directly

File: pipelines/google_ads/__init__.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+            "developer_token": dev_token,
+            "refresh_token": credentials.refresh_token,
+            "client_id": credentials.client_id,
+            "client_secret": credentials.client_secret
+        }
+        client = GoogleAdsClient.load_from_dict(config_dict=credentials)
+    # use service account to authenticate if not using OAuth2.0
+    else:
+        pass
```
Comment (rudolfix) [inline pipelines/google_ads/__init__.py:47]: not yet done I assume? both credentials have `to_google_credentials` so you can just do it at the end?

File: pipelines/google_ads/__init__.py [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+               dev_token: str = dlt.secrets.value):
+    """
+    Loads default tables for google ads in the database.
+    :param credentials:
+    :param dev_token:
+    :return:
+    """
+    # generate access token for credentials if we are using OAuth2.0
+    if isinstance(credentials, GoogleAnalyticsCredentialsOAuth):
+        credentials.auth("https://www.googleapis.com/auth/analytics.readonly")
```
Comment (rudolfix) [inline pipelines/google_ads/__init__.py:73]: wrong scope I assume?

File: pipelines/google_ads/generate_user_credentials.py [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+import socket
+import sys
+from urllib.parse import unquote
+
+# If using Web flow, the redirect URL must match exactly what’s configured in GCP for
+# the OAuth client.  If using Desktop flow, the redirect must be a localhost URL and
+# is not explicitly set in GCP.
+from google_auth_oauthlib.flow import Flow
+
+_SCOPE = "https://www.googleapis.com/auth/adwords"
```
Comment (rudolfix) [inline pipelines/google_ads/generate_user_credentials.py:32]: why are we using the web flow not a desktop flow? we had nice script for google analytics. any reason?

**Review by rudolfix — CHANGES_REQUESTED** [review]
Are you planning to implement incremental?

**Comment by rudolfix** [comment]
here's the feedback to the current state:
## Authentication
A developer token is needed on top of regular google authentication. Only OAuth is currently implemented as a way of authenticating. The OAuthCredential Class in dlt core might need to be better
___
## Pipeline Script and Sources
These are setup currently. There are 2 sources: 1 for user queries and 1 for default tables. User queries are saved in config or given as input for the user queries source.
## Helpers and testing
Helpers are set up in helpers/data_processing.py to process the responses from the google api client. The problem being that responses are mostly Objects which have no implemented to_dict() or to_json() method. Some testing sample functions are also set up.
___
## Resources
For both the resources of the default and user query tables major changes are needed. A query is needed to get this data and in Google Ads Query language “Select *” is not a thing, so column names need to be included in the config which will make the whole thing bloated.

File: pipelines/google_ads/__init__.py (hunk 1 above)
Comment (sh-rp) [reply pipelines/google_ads/__init__.py:4]: done

File: pipelines/google_ads/__init__.py (hunk 2 above)
Comment (sh-rp) [reply pipelines/google_ads/__init__.py:38]: we need a dict here to inject the developer key, somehow google ads works a bit different. we can discuss

File: pipelines/google_ads/__init__.py (hunk 4 above)
Comment (sh-rp) [reply pipelines/google_ads/__init__.py:73]: fixed

**Pull request closed without merging**